File: java/duplicateAndAddNewSegment.java
import java.io.FileInputStream;
import java.io.IOException;

import java.util.ArrayList;
import java.util.Map;
import java.util.Properties;

import org.json.JSONException;
import org.json.JSONObject;

import Utils.Request;


public class duplicateAndAddNewSegment
{
    private static JSONObject getNewActionDuplicateJson(JSONObject actionDuplicateJson,
                                                        int[] segmentIds) throws JSONException
    {
        //Modification de la nouvelle action
        JSONObject segments = new JSONObject();
        segments.put("selected", segmentIds);

        JSONObject scheduler = new JSONObject();
        scheduler.put("type", "asap");	//Envoi : immediat = 'asap' / Date = 'scheduled'
        //scheduler.put("startDate", "2015-07-27T08:15:00Z");	//Si type = 'scheduled' sinon a enlever
        scheduler.put("segments", segments);

        actionDuplicateJson.put("scheduler", scheduler);

        return (actionDuplicateJson);
    }

    private static JSONObject getActionValidation(int[] segmentTestIds) throws JSONException
    {
        JSONObject actionValidationJson = new JSONObject();
        actionValidationJson.put("fortest", true);	//Phase de test
        actionValidationJson.put("campaignAnalyser", false);	//Campaign Analyzer : 'true' = oui / 'false' = non
        actionValidationJson.put("testSegments", segmentTestIds);	//Les Ids des differents segments de tests
        actionValidationJson.put("mediaForTest", JSONObject.NULL);	//Rediriger tous les tests vers une seule adresse ('JSONObject.NULL' pour aucune valeur)
        actionValidationJson.put("textandHtml", false);	//Envoyer la version texte et la version html : 'true' = oui / 'false' = non
        actionValidationJson.put("comments", JSONObject.NULL);	//Commentaire ('JSONObject.NULL' pour aucuns commentaires)

        //On affiche le Json
        System.out.print(actionValidationJson + "\n");

        return (actionValidationJson);
    }

    public static ArrayList duplicateAction(String urlBase,
                                            String actionId, String xKey) throws IOException, JSONException
    {
        //Nouvelle url
        String url = urlBase + "actions/" + actionId + "/duplication";

        JSONObject data = new JSONObject();

        System.out.print("Duplicating the Action\n\n");

        //On duplique a l'identique
        Map<String, String> resp = Request.connection(url, xKey, data, "POST");

        data = new JSONObject(resp.get("reply"));

        if (resp.get("code").equals("200")) {
            System.out.print("The action duplicate \"" + data.getString("name") + "\" has been created :\n");

            System.out.print(resp.get("reply") + "\n\n");
        }
        else {
            // Affichage de l'erreur
            System.out.print("Error : " + resp.get("code") + " " + resp.get("mess"));
            return null;
        }

        //On recupere l'id de la nouvelle action
        String actionDuplicateId = data.getString("id");

        ArrayList result = new ArrayList();
        result.add(resp);
        result.add(actionDuplicateId);
        result.add(data);

        return (result);
    }

    public static int updateActionDuplicate(String urlBase, String actionDuplicateId,
                                            JSONObject actionDuplicateJson, int[] segmentIds,
                                            String xKey) throws IOException, JSONException
    {
        Map<String, String> resp;

        //Nouvelle url
        String url = urlBase + "actions/" + actionDuplicateId;

        System.out.print("Adding New Segment\n\n");

        //On duplique a l'identique
        resp = Request.connection(url, xKey, getNewActionDuplicateJson(actionDuplicateJson, segmentIds), "PUT");

        if (resp.get("code").equals("200")) {
            System.out.print("The duplicated action \"" + new JSONObject(resp.get("reply")).getString("name") +
                    "\" is update:\n");
            System.out.print(resp.get("reply") + "\n\n");
        }
        else {
            // Affichage de l'erreur
            System.out.print("Error : " + resp.get("code") + " " + resp.get("mess") + "\n");
            return -1;
        }

        return 0;
    }

    public static int testActionDuplicate(String urlBase, String actionDuplicateId,
                                          String xKey, int[] segmentTestIds) throws InterruptedException, IOException, JSONException
    {
        //Nouvelle url
        String url = urlBase + "actions/" + actionDuplicateId + "/validation";

        System.out.print("Testing the duplicated Action\n\n");

        //Lancement de la connexion
        Request.connection(url, xKey, getActionValidation(segmentTestIds), "POST");

        int actionState = Request.waitForState(actionDuplicateId, xKey, urlBase);

        if (actionState != 38) {
            //Affichage de l'erreur
            if (actionState == 20)
                System.out.print("Error : the test failed.");
            else if (actionState == 10)
                System.out.print("Error : check the campaign in the Backoffice.");
            return -1;
        }
        else {
            //La phase de test a reussi
            System.out.print("The action has been tested.\n\n");
        }
        return 0;
    }

    public static void main(String[] args) throws Exception
    {
        Properties properties   = new Properties();

        try {
          properties.load(new FileInputStream("config.properties"));
        } catch (IOException e) {
          System.out.print(e.getMessage());
        }

        String xKey = properties.getProperty("xKey");
        String baseUrl = properties.getProperty("url");

        int[] segmentIds = {01234};	//Id du segment a modifier
        int[] segmentTestIds = {01234};	//Id du segment de test a modifier
        String actionId = "000ABC";	//Id de l'action a dupliquer

        ArrayList result;


        if ((result = duplicateAction(baseUrl, actionId, xKey)) == null)
            return;
        String actionDuplicateId = (String) result.get(1);
        JSONObject actionDuplicateJson = (JSONObject) result.get(2);

        if (updateActionDuplicate(baseUrl, actionDuplicateId, actionDuplicateJson, segmentIds, xKey) == -1)
            return;

        if (testActionDuplicate(baseUrl, actionDuplicateId, xKey, segmentTestIds) == -1)
            return;

        return;
    }
}

File: php/duplicateAndAddNewSegment.php
<?php

/**
 * Ce script permet de dupliquer une action et d'ajouter à un ségment
 * existant ou non, si il n'existe pas il est créé.
 *
 * @package cookbook
 */

require 'utils.php';

/**
 * Variable contenant les configurations pour se connecter à l'API
 *
 * @var array
 */
$configs = parse_ini_file("config.ini");
/**
 * Id de l'action a dupliquer
 *
 * @var string
 */
$idAction = "XXXXXX";
/**
 * Id du segment de test pour permettre de valider l'action dupliquée
 *
 * @var array
 */
$idTestSegment = array(01234);
/**
 * Id du segment à ajouter à la nouvelle action.
 *
 * @var array
 */
$idSelectSegment = array(01234);
/**
 * Creation du JSON contenant les informations
 * (pour plus de détails : https://backoffice.mailperformance.com/doc/#api-Action-ValidateAsync)
 *
 * @var array
 */
$arr = array(
  'fortest'           => true,	// Phase de test
  'campaignAnalyser'  => false,
  'testSegments'      => $idTestSegment,
  'mediaForTest'      => null,	// Tests vers une adresse
  'textandHtml'       => false,
  'comments'          => null);

// Duplication de l'action

$url = $configs['url'] . "actions/" . $idAction . "/duplication";
$con = connect($url, $configs['xKey'], null, "POST");

if ($con['info']['http_code'] == 200) {
  // On affiche la cible
  echo "The action duplicate ( " . json_decode($con['result'], TRUE)['name'] . " ) is created.\n";
  echo $con['result'] . "\n\n";

  // On recupere l'id de l'action
  $tab = json_decode($con['result'], TRUE);
  $idNewAction = $tab['id'];

  $data = json_decode($con['result'], TRUE);

  // On ajoute le nouveau Segment à l'action
  $data['scheduler']['segments'] = array(
                                          'selected'      => $idSelectSegment
                                        );
  $dataJson = json_encode($data);
  $url = $configs['url'] . "actions/" . $idNewAction;
  $con = connect($url, $configs['xKey'], $dataJson, "PUT");
  if ($con['info']['http_code'] != 200) {
    echo "\nError : " . $con['info']['http_code'] . "\n";
    return;
  }

  // On affiche le Json
  $message = json_encode($arr);
  echo $message . "\n";

  // Connexion
  $url = $configs['url'] . "actions/" . $idNewAction . "/validation";
  $con = connect($url, $configs['xKey'], $message, 'POST');
  $result = $con['result'];
  $info = $con['info'];

  // On attend que la phase de test soit terminée
  $actionState = waitForState($idNewAction, $configs['url'],$configs['xKey']);

  // On verifie les reponses
  if ($info['http_code'] != 204 || $actionState != 38) {
    if ($actionState == 20)
      echo 'Error : the test failed';
    else if ($actionState == 10)
      echo 'Error : check the campaign in the Backoffice.';
    else
      echo 'Error : ' . $info['http_code'];
  }
  else {
    // La phase de test a reussi
    echo "The action " .
         json_decode($result, TRUE)['name'] .
         " has been tested.\n\n";
  }
}
else {
  echo "Error request 'POST' duplication : " . $info['http_code'];
}

?>
